Path creation

dwatchguard opened the pull request:
Firstly, I changed the name of PathMap to CostMap because it made more sense and if I want to make a map of paths later, it would be confusing.

Other than that, the main addition was 2 classes: PathBase and MovementPath

PathBase has simple push_back,  clear, etc.. methods and was created so I have  a base for classes like AngularCellPath in the future

MovementPath was made specifically for moving Units through the Map, and MoveHelper now has Methods for doing exactly that.

Issues you will most likely have:
PathBase should take a template allowing for GridCell or const GridCell, but I am having weird `typedef` issues with that so I left that out for now

`MovementPath(CostMap& map)` should maybe be replaced with  `MovementPath GridMap::getShortestPath(...)`

Resolve #26

dwatchguard committed:
Initial CellPath Commit

dwatchguard committed:
Merge

dwatchguard committed:
Fixed Merge

dwatchguard committed:
Improved merge

dwatchguard committed:
Logic moved to PathBase

dwatchguard committed:
Merge again from map_refactor

dwatchguard committed:
Implemented findShortestPath for Units
Implemented a simple walk method

dwatchguard committed:
Renamed PathMap to CostMap
Added more list fucnctionality to PathBase

dwatchguard committed:
GridMap: Added more parameter options for getting Unit CostMaps

MovementPath now has a Unit&, and can be appended using +

MoveHelper can now find the most desireable MovementPath to a destination if given another Path

dwatchguard committed:
Gave MovePath a Head, and cleaned up +operation

dwatchguard committed:
Improved MovementPath::push_back() to better handle cycles

dwatchguard committed:
Fixed MovementPath.front()

dwatchguard committed:
PathBase now uses GridCell::ConstRef and has a _head

dwatchguard committed:
Merge branch 'master' into path_creation

dwatchguard committed:
Cleaned up #includes and forward declares

SilensAngelusNex commented on Project1/GridMap.h:
```diff
@@ -21,15 +21,18 @@ class GridMap :	public Grid {
+	CostMap getShortestPathsMap(ID start, int max_move, MobilitySet mobility);
+	CostMap getShortestPathsMap(ID start, int max_move, MobilitySet mobility, bool intangible);
+	CostMap getShortestPathsMap(const Unit& unit);
+	CostMap getShortestPathsMap(const Unit& unit, ID start);
```
Why the ID? Doesn't the unit have to start on the space it's on?

dwatchguard replied:
They don't yet

dwatchguard replied:
It also says, if the Unit _was_ somewhere else, where could they go. It can be useful

SilensAngelusNex commented on Project1/MoveHelper.cpp:
```diff
@@ -142,3 +142,52 @@ std::vector<ID> MoveHelper::getOtherAlliedCellIDs(const Unit& unit) const{
 	////////////////////////////////////////////////////////////////////////////////
 
+MovementPath MoveHelper::getShortestPath(const Unit& unit, ID destination) {
+	return MovementPath(unit, _map.getShortestPathsMap(unit), _map[destination]);
```
What if the unit can't reach the destination? Does that mean it will return a path that's longer than the unit's movement or would it make more sense to return `optional<MovementPath>()`?

dwatchguard replied:
That currently throws an out of bounds error, I think an Expects clause should do enough

SilensAngelusNex commented on Project1/MoveHelper.cpp:
```diff
@@ -142,3 +142,52 @@ std::vector<ID> MoveHelper::getOtherAlliedCellIDs(const Unit& unit) const{
+}
+
+//TODO(Torrey): optomize this
+MovementPath& MoveHelper::reRoutePath(MovementPath& path, ID destination) {
```
When does this get called? It isn't clear to me what it's doing. Is destination the new destination that it's getting rerouted to?

dwatchguard replied:
This is complicated, I'll explain it tonight.

dwatchguard replied:
Basically, when you click on a unit, it shows every space it can reach, and will make a path with only a head.  You then use this algorithm to 'make' the paths. What it essentially does is try to use as much of the current path as possible when you try to append cells to the path.

1) When you hover over a tile, if it unreachable, the path remains unchanged.

2) If it is adjacent, reachable, and directly adding it to our current path does NOT exceed our max move, we push_back() that tile.

3) If it is not adjacent to the end of our current path, but can still reach it using our current path as a base, it appends the shortest path to that tile from the back of our current path to the end of our path.

4) If that tile is already in our path, we trim back to it, NO LOOPS.

5) And finally, If the tile is reachable, but there is no way for our current path to reach it without exceeding our max move, we pop off the back cell and attempt rule 4 (above). If that doesn't work, we pop again, and again,and again... until it eventually HAS TO, because we already determined there was SOME path to get there.

SilensAngelusNex commented on Project1/MoveHelper.cpp:
```diff
@@ -142,3 +142,52 @@ std::vector<ID> MoveHelper::getOtherAlliedCellIDs(const Unit& unit) const{
+		}
+		path.pop_back();
+	}
+	Expects(false);
```
This is a postcondition, so use `Ensures(false)`

dwatchguard replied:
kk

SilensAngelusNex commented on Project1/PathBase.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+void PathBase::trimPath(ID end_id) {
+	while (!_path.empty() && back()._id != end_id) {
```
I think these should be
```
Expects(contains(end_id));
while (back()._id != end_id) {
	pop_back();
}
```
because giving an id/gridcell not in the path has to be programmer error.

dwatchguard replied:
It was coded that way so that it would stop when it hit the head, which wasn't in the list

SilensAngelusNex commented on Project1/PathBase.h:
```diff
@@ -0,0 +1,53 @@
+#include "GridCell.h"
+
+
+class PathBase {
```
I would name these GridPath and UnitPath I think.

SilensAngelusNex commented on Project1/MoveHelper.cpp:
```diff
@@ -142,3 +142,52 @@ std::vector<ID> MoveHelper::getOtherAlliedCellIDs(const Unit& unit) const{
+			//Add logic for that
+			return;
+		}
+		if (!_map.hasUnit(pair.second)) { //Skip over units we can move through
```
We're going to need to add something to let a the units know they stepped on someone/someone stepped on them and let the moving unit know that it stepped on a tile.

Somewhat unrelatedly, what happens when someone with Pass walks through someone stealthed?

dwatchguard replied:
After each step, I guess we would let every one passed through know. And stealth makes everything hard

SilensAngelusNex reviewed (commented):
Everything except for reroute makes pretty good sense to me. Just a few small nits.

dwatchguard committed:
PathBase now takes template arguments
Rename MovementPath to UnitPath
`class UnitPath : pubilc PathBase<const GridCell>`
PathBase heads are now references, and cannot no longer be changed.
Removed `constCostMap getShortestPathsMap(const Unit& unit, ID start) const;`

Fixed compiler warnings in BitsetIterator and UtilTests

dwatchguard committed:
Rename `reRoutePath()` to `routePathTo()`
add comment to explain its purpose

dwatchguard committed:
` PathBase::back()` now grabs the head if _path is empty, and has a non const variant.
